### PR TITLE
add wrapers and extra method handlers

### PR DESCRIPTION
**Pull request by nabam**

### CODE REVIEW

File: pkg/httpclient/client.go [hunk 1]
```diff
@@ -9,7 +10,29 @@ import (
 )
 
 var (
-	httpClient *http.Client = &http.Client{
+	errNoClient          = errors.New("HTTP client is not initialized, call InitClient()")
```
Comment (ronanbarrett) [inline pkg/httpclient/client.go]: Move down so we have just one `var` section.

File: pkg/httpclient/client.go [hunk 2]
```diff
@@ -9,7 +10,29 @@ import (
...
+// InitClient should be run only once when starting the server
+func InitClient(c *http.Client) error {
+	if httpClient != nil {
+		return errClientInitialized
+	}
+	httpClient = c
+	return nil
+}
+
+// InitClientWithDefaults should be run only once when starting the server
```
Comment (ronanbarrett) [inline pkg/httpclient/client.go]: The comment doesn't actually say what this function does.

File: pkg/httpclient/client.go [hunk 3]
```diff
@@ -9,7 +10,29 @@ import (
...
-	httpClient *http.Client = &http.Client{
+	errNoClient          = errors.New("HTTP client is not initialized, call InitClient()")
+	errClientInitialized = errors.New("HTTP client is already initialized")
+)
+
+var (
+	httpClient *http.Client
+)
+
+// InitClient should be run only once when starting the server
```
Comment (ronanbarrett) [inline pkg/httpclient/client.go]: The comment doesn't actually say what this function does.

File: pkg/httpclient/client.go [hunk 4]
```diff
@@ -18,33 +41,94 @@ var (
...
+	InternalPut(url, contentType string, body io.Reader) (err error)
+	InternalDo(request *http.Request) (resp *http.Response, err error)
 }
 
-// CreateScopedClient creates ScopedClient with the one and only HTTP Client
+// GetScopedClient returns ScopedClient with the one and only HTTP Client
 // and given downstream request scope
-func CreateScopedClient(c *gin.Context) Client {
-	return &ScopedClient{HTTPClient: httpClient, RequestContext: c}
+func GetScopedClient(c *gin.Context) (Client, error) {
```
Comment (ronanbarrett) [inline pkg/httpclient/client.go]: "GetScopedClient returns ScopedClient with the one and only HTTP Client and given downstream request scope" -> "GetScopedClient returns ScopedClient with the one and only HTTP Client preconfigured with the provided downstream request scope"

File: pkg/httpclient/client.go [hunk 5]
```diff
@@ -18,33 +41,94 @@ var (
...
+// Get invokes HTTP GET
+func Get(url string) (resp *http.Response, err error) {
+	if httpClient == nil {
+		return nil, errNoClient
+	}
+	return httpClient.Get(url)
+}
+
+// Head invokes HTTP HEAD
+func Head(url string) (resp *http.Response, err error) {
```
Comment (ronanbarrett) [inline pkg/httpclient/client.go]: Isn't this missing from the interface? ^^

File: pkg/httpclient/scoped_client.go [hunk 6]
```diff
@@ -225,3 +194,78 @@ func (sc *ScopedClient) delete(url string, internal bool) error {
...
+
+	resp, err := sc.do(req, internal)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	// PUT request expects "204 No Content" response
+	err = sc.checkStatus(http.StatusNoContent, resp)
```
Comment (ronanbarrett) [inline pkg/httpclient/scoped_client.go]: You need to be super careful that when this is used that all our services play by this rule? I hope they do but I cannot say for sure. Same comment for Delete.

File: pkg/httpclient/scoped_client_test.go [hunk 7]
```diff
@@ -252,18 +337,20 @@ func TestCheckServerStatusWith4xx(t *testing.T) {
 
 // Helpers below
 func newTestRequest() (client Client, context *gin.Context, recorder *httptest.ResponseRecorder) {
+	_ = InitClientWithDefaults()
```
Comment (ronanbarrett) [inline pkg/httpclient/scoped_client_test.go]: `_ = InitClientWithDefaults()` -> `InitClientWithDefaults()`

File: pkg/httpclient/client.go (hunk 1 above)
Comment (nabam) [reply pkg/httpclient/client.go]: I think it's quite conventional to have multiple var sections:
https://golang.org/doc/effective_go.html#commentary

File: pkg/httpclient/scoped_client_test.go (hunk 7 above)
Comment (nabam) [reply pkg/httpclient/scoped_client_test.go]: I like this more as it explicitly shows that code ignores the error

File: pkg/httpclient/scoped_client.go (hunk 6 above)
Comment (nabam) [reply pkg/httpclient/scoped_client.go]: Well, this is how it works now. I agree that it's not safe, should I change it?

Comment (nabam) [reply pkg/httpclient/scoped_client.go]: all 2xx codes should be treated as valid in my opinion

File: pkg/httpclient/client.go (hunk 5 above)
Comment (nabam) [reply pkg/httpclient/client.go]: It is, but it's not an interface implementation, it's wrapper functions around net/http.client

File: pkg/httpclient/client_test.go [hunk 8]
```diff
@@ -7,15 +7,23 @@ import (
 	"testing"
 )
 
-func TestGetClient(t *testing.T) {
+func TestGetScopedClient(t *testing.T) {
+	httpClient = nil
+	_, e := GetScopedClient(nil)
```
Comment (ronanbarrett) [inline pkg/httpclient/client_test.go]: `e` -> `err` everywhere for readability

File: pkg/httpclient/client_test.go [hunk 9]
```diff
@@ -7,15 +7,23 @@ import (
 	"testing"
 )
 
-func TestGetClient(t *testing.T) {
+func TestGetScopedClient(t *testing.T) {
+	httpClient = nil
```
Comment (ronanbarrett) [inline pkg/httpclient/client_test.go]: A comment is warranted here to say you are ensuring client is nil before running test i.e. precondition

File: pkg/httpclient/client_test.go [hunk 10]
```diff
@@ -7,15 +7,23 @@ import (
...
+	client, _ := GetScopedClient(context)
+	require.IsType(t, &ScopedClient{}, client, "GetScopedClient should return scoped client")
+	require.Equal(t, context, client.(*ScopedClient).requestContext, "ScopedClient should refer to given context")
 }
 
 func TestInitClient(t *testing.T) {
-	d := httpClient
-	Configure(&http.Client{})
-	require.NotEqual(t, d, HttpClient(), "Configure should override default client")
+	httpClient = nil
```
Comment (ronanbarrett) [inline pkg/httpclient/client_test.go]: Same comment as above.

File: pkg/httpclient/client_test.go [hunk 11]
```diff
@@ -7,15 +7,23 @@ import (
...
+	e := InitClientWithDefaults()
+	require.Nil(t, e, "Init should succeed first time")
```
Comment (ronanbarrett) [inline pkg/httpclient/client_test.go]: Use `require.NoError`

File: pkg/httpclient/client_test.go [hunk 12]
```diff
@@ -7,15 +7,23 @@ import (
...
+	c := &http.Client{}
+	e = InitClient(c)
+	require.NotNil(t, e, "Init should not be allowed twice")
```
Comment (ronanbarrett) [inline pkg/httpclient/client_test.go]: Use `require.Error`

File: pkg/httpclient/scoped_client_test.go [hunk 13]
```diff
@@ -20,122 +20,115 @@ var (
...
+						for _, h := range PassthroughHeadersInternal {
+							require.Equal(t, "", r.Header.Get(h), tc.description+": internal headers should not be propagated upstream")
+						}
+					}
+				}})
+			defer server.Close()
+
+			client, _, _ := newTestRequest()
+			body, err := tc.call(server.URL, client, requestBody)
+			require.Nil(t, err, tc.description+": successful call should not return error")
```
Comment (ronanbarrett) [inline pkg/httpclient/scoped_client_test.go]: use `NoError`

File: pkg/httpclient/scoped_client_test.go [hunk 14]
```diff
@@ -20,122 +20,115 @@ var (
...
+				Body:     []byte{},
+				Callback: nil,
+			})
+			defer server.Close()
+
+			client, context, recorder := newTestRequest()
+			body, err := tc.call(server.URL, client, requestBody)
+
+			require.Nil(t, body, tc.description+": call with issue should return no body")
+			require.NotNil(t, err, tc.description+": call with issue should return error")
```
Comment (ronanbarrett) [inline pkg/httpclient/scoped_client_test.go]: use `require.Error`

File: pkg/httpclient/scoped_client_test.go [hunk 15]
```diff
@@ -20,122 +20,115 @@ var (
...
+			require.True(t, context.IsAborted(), tc.description+": call with issue should abort original request")
+			require.Equal(t, http.StatusInternalServerError, recorder.Code, tc.description+": call with issue should send error status downstream")
+			require.Equal(t, testServerErrorBody, recorder.Body.Bytes(), tc.description+": call with issue should send error downstream")
```
Comment (ronanbarrett) [inline pkg/httpclient/scoped_client_test.go]: "call with issue should send error downstream" -> "call with issue should send error body downstream"

File: pkg/httpclient/scoped_client_test.go [hunk 16]
```diff
@@ -20,122 +20,115 @@ var (
...
-	server := httptest.NewServer(&testHTTPServerHandler{
-		Status:   http.StatusNoContent,
-		Body:     []byte{},
-		Callback: nil,
-	})
-	defer server.Close()
-
-	_test := func(err error) {
-		require.Nil(t, err, "Successful Delete should not return error")
+func TestMethodShortcuts(t *testing.T) {
```
Comment (ronanbarrett) [inline pkg/httpclient/scoped_client_test.go]: Use descriptive test name

File: pkg/httpclient/scoped_client_test.go [hunk 17]
```diff
@@ -145,72 +138,22 @@ func TestDo(t *testing.T) {
...
-		headers[h] = fmt.Sprintf("value of header %s", h)
-	}
-	for _, h := range PassthroughHeadersInternal {
-		headers[h] = fmt.Sprintf("value of internal header %s", h)
-	}
-
-	client, context, _ := newTestRequest()
-	for h, v := range headers {
-		context.Request.Header.Set(h, v)
+		require.Nil(t, err, "Do should not return error")
```
Comment (ronanbarrett) [inline pkg/httpclient/scoped_client_test.go]: NoError

**Pull request merged**